[문자열 계산기] 제출합니다~

chunghyeon-kimm opened the pull request:
미흡한 부분 있으면 피드백 부탁드립니다 (_ _)

chunghyeon-kimm committed:
, : 구분자를 이용한 계산 구현

chunghyeon-kimm committed:
커스텀 구분자 구현

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,42 @@
+        if(text == null | "".equals(text)) {
+            return 0;
+        }
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
+        if(m.find()) {
+            return seperateAndCalculate(m.group(2), m.group(1));
+        }
```
해당 코드를 보고 어떻게 동작할지에 대해서는 이해가 되지만,
해당 로직이 의미하는 바가 무엇인지를 명시적으로 드러내면 코드를 읽는 사람에게 더 도움이 될 것 같아요.

어떻게 의도를 드러낼 수 있을까요?

chunghyeon-kimm replied:
분리된 문자열들을 적절한 이름을 가진 변수로 지정했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,42 @@
+        if(m.find()) {
+            return seperateAndCalculate(m.group(2), m.group(1));
+        }
+        return seperateAndCalculate(text, ",|:");
```
`",|:"`가 의미하는 바가 무엇일까요?
이런 의미가 불분명한 값을 매직넘버라고 부릅니다.
해당 값이 무엇을 의미하는지 코드를 읽는 사람이 명확하게 받아드릴 수 있도록 하려면 어떻게 하면 좋을까요?
매직넘버에 대해서 한 번 찾아보셔도 좋을 것 같네요 :)

chunghyeon-kimm replied:
매직넘버의 의미를 파악하기 힘들고 코드 수정시 혼동을 줄 수 있다는 것을 알게 되었습니다.
delimiter라는 이름의 상수 String 배열을 선언해서 의미를 알 수 있게 수정했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,42 @@
+        String[] numbers = text.split(delimeter);
+        int[] intArr = new int[numbers.length];
+
+        for(int i=0; i<numbers.length; i++) {
+            intArr[i] = Integer.parseInt(numbers[i]);
+            if(intArr[i] < 0) {
+                throw new RuntimeException();
+            }
+        }
```
`프로그래밍 요구사항 1. indent는 1을 넘지 않는다.`를 반영해주세요!

chunghyeon-kimm replied:
수정했습니다.

jjj0611 commented on src/main/java/calculator/test.java:
```diff
@@ -0,0 +1,16 @@
+
+public class test {
+    public static void main(String[] args) {
+        String input = "//;\n1;2;3";
+        System.out.println(input);
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(input);
+        if(m.find()) {
+            System.out.println(m.group(1));
+            System.out.println(m.group(2));
+        }
```
학습 테스트를 작성해주셨네요.
학습 테스트는 /src/main이 아닌 /src/test로 들어가는 것이 좋을 것 같습니다.
src/main package에 main method로서 학습테스트가 들어갈 경우 어떤 점이 좋지 않을지 고민해보면 좋을 것 같아요.

또한, 개인적으로 학습테스트는 다른 사람들이 보았을 때도 학습테스트의 의도가 명확하게 드러나면 좋을 것 같습니다.
의도가 명확한 테스트가 되기 위해서 어떻게 해야할지, 어떠한 테스트가 좋은 테스트인지 고민해보면 좋을 것 같아요!

chunghyeon-kimm replied:
이런 코드를 '학습 테스트'라고 부른다는 것을 알게 되었습니다.
src/main package에 main method로서 학습테스트가 들어갈 경우 다른사람이 봤을 때 테스트 목적의 코드인지 아닌지 알기 어려울 것 같습니다.
junit을 사용해서 테스트 결과를 직관적으로 파악할 수 있도록 수정했습니다.

jjj0611 commented on src/pom.xml:
```diff
@@ -0,0 +1,52 @@
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>8</source>
+                    <target>8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-engine -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>5.7.0</version>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.0</version>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.assertj/assertj-core -->
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.16.1</version>
+            <scope>test</scope>
+        </dependency>
+
+
+    </dependencies>
+
+    
+</project>
```
pom.xml은 어떤 의도로 추가하셨을까요?

chunghyeon-kimm replied:
clone해온 프로젝트를 open할 때 실수로 프로젝트 하위의 src폴더로 open했습니다.
의존성 관리도구가 없는 줄 알고 maven을 적용하고 pom.xml에 의존성을 추가했습니다.

삭제했습니다 :)

jjj0611 reviewed (changes requested):
충현님 구현해주신 내용 잘 봤습니다.
전체적으로 깔끔하게 잘 작성해주셨네요 👍

프로그래밍 요구사항이 지켜지지 않은 부분이나 수정되면 좋을 것 같은 내용에 대해서
코멘트 드렸으니 확인 부탁드려요.

또한 저장소는 다른 사람과 함께 사용하는 공간이니(일반적 협업 시에)
.class 파일 같은 경우는 commit에 올라가지 않도록
주의해서 commit을 하는 습관을 들이면 좋을 것 같아요 😃

chunghyeon-kimm committed:
StringCalculator Refactoring

chunghyeon-kimm committed:
second Refactoring, move and rename test.java

delimiter String 상수 배열 선언 후 입력

학습테스트 이동, 파일이름 변경, 주석 추가

chunghyeon-kimm committed:
학습테스트 수정

학습테스트에 junit 적용

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+        System.out.println(splitAndSum(text));
+    }
+
+    protected static final String[] delimiter = {",", ":"};
```
protected로 선언하신 이유가 있을까요?

프로그래밍 언어에는 흔히 사용하는 컨벤션이라는 개념이 존재합니다.
JAVA에서도 자주 사용하는 컨벤션이 있는데, static 변수에서의 네이밍 컨벤션이 어떤지 한 번 찾아보면 좋을 것 같아요.

이건 약간 사족이긴 한데, delimiter라는 변수명은
- 해당 변수는 배열인데 단수인 것 같은 이름
- custom delimiter(사용자가 넣은 구분자)도 delimiter인데, 너무 포괄적 의미
라는 생각이 들긴 합니다!

앞으로도 계속해서 고민하시겠지만, 프로그래밍에서 이름 짓기는 다른 사람의 코드 이해를 도와주는 아주 중요한 요소입니다 :)
네이밍에는 많은 고민을 하는 노력을 들이면 좋을 것 같아요!

chunghyeon-kimm replied:
sonarLint 플러그인 메세지를 보고 protected로 선언했는데 제가 메세지를 잘못 본 것 같습니다. private으로 변경했습니다.
상수의 이름은 대문자와 언더스코어를 이용해서 지어준다는 것을 다시한번 상기했습니다.
delimiter라는 변수명은 지적해주신대로 잘 맞지 않는 것 같아서 DEFAULT_DELIMITERS로 변경했습니다.

앞으로 네이밍에 더 신경쓰겠습니다. 좋은 지적 감사합니다 :)

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+        if(text == null | "".equals(text)) {
+            return 0;
+        }
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
```
요 친구도 일종의 매직넘버라고 볼 수 있지 않을까요?

chunghyeon-kimm replied:
DESIGNATED_PATTERN이라는 상수로 지정했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+            String content = m.group(2);
+            return seperateAndCalculate(content, customDelimiter);
+        }
+        return seperateAndCalculate(text, delimiter[0] + "|" + delimiter[1]);
```
매직넘버에 대한 분리는 되었지만, ",|;"로 존재할 때보다 가독성이 떨어지는 것 같아요.
문자열 계산기의 정책 중 하나가 콤마(,) 또는 세미콜론(;)으로 숫자를 구분하여 더하는거니
",|;"를 묶어서 의미를 부여해줄 수 있을 것 같은데 어떻게 생각하시나요? :)

chunghyeon-kimm replied:
구분자를 , 혹은 : 2가지로 지정해주었기 때문에 묶어서 의미를 부여해주는 게 확실히 더 나은 방법인 것 같습니다.
수정했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+        int[] intArr = new int[numbers.length];
+
+        for(int i=0; i<numbers.length; i++) {
+            intArr[i] = Integer.parseInt(numbers[i]);
```
numbers[i] 가 String이면 프로그램은 어떻게 될까요?

chunghyeon-kimm replied:
NumberFormatException이 발생합니다.
매개변수 유효성검사 메서드를 추가해서 Integer.parseInt() 메서드의 매개변수로 숫자가 아닌 문자열이 들어가는 상황을 제외했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+            intArr[i] = Integer.parseInt(numbers[i]);
+        }
+
+        Arrays.stream(intArr).forEach(StringCalculator::negativeCheck);
```
Stream을 써보셨네요! Java8부터 제공하는 Stream 기능은 매우 막강하지만,
조금은 천천히 도입해보는 것도 좋을 것 같아요!

현재 존재하는 stream 사용 부분을 모두 for문으로 바꾸어볼까요?

chunghyeon-kimm replied:
수정했습니다.

jjj0611 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static void negativeCheck(int input) throws RuntimeException{
+        if(input < 0) {
+            throw new RuntimeException();
+        }
```
사용자가 "3,5,8,-1"을 입력하면 예외가 나고 프로그램이 종료되겠군요.
하지만 위와 같이 예외가 던져지면 사용자는 "어? 프로그램이 왜 종료됐지?"라는 의문을 해소할 수 없을 것 같아요.

좀 더 유저 친화적인 프로그램을 만드려면 어떻게 하는게 좋을까요?

chunghyeon-kimm replied:
생성되는 RuntimeException 객체에 메세지를 추가했습니다.

jjj0611 commented on src/test/java/calculator/MatcherStudyTest.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.regex.Pattern;
+
+
+public class MatcherStudyTest {
```
학습테스트를 작성함으로써 내가 사용하려는 기능이 제대로 동작하는지를 보장할 수 있어서 학습테스트 작성은 매우 좋은 습관입니다 :)
assertThat 구문을 사용하여 테스트를 완료하니 값에 대한 검증까지 이루어져 m.find()와 m.group()의 기능을 안전하게 사용할 수 있습니다 :)

chunghyeon-kimm replied:
네. 앞으로도 학습테스트를 적극적으로 활용하겠습니다 :)

jjj0611 reviewed (commented):
충현 선생님 리뷰 잘 반영해 주셨네요 :)

조금 더 고민해보면 좋을만한 부분에 대해서 코멘트 남겼으니 한 번 확인해주세요!
일단 Request changes는 하지 않고, 조금 더 수정해보겠다 하시면 수정하시고 다시 요청 주시고,
다음 단계를 진행하면서 천천히 반영해나가보겠다 하시면 말씀주시면 merge 해드리겠습니다!

chunghyeon-kimm committed:
2차 Refactoring

상수 접근제어자 private으로 변경
상수 이름 대문자로 변경
Pattern 상수 지정
NumberFormatException 예외처리 추가
Stream -> for문으로 변경

chunghyeon-kimm committed:
예외처리 제거

chunghyeon-kimm committed:
정규표현식을 이용한 유효성검사 추가

chunghyeon-kimm committed:
필요없는 배열 제거, 변수이름 변경

chunghyeon-kimm committed:
향상된 for문으로 변경

chunghyeon-kimm committed:
접근제어자 수정

jjj0611 commented:
충현 선생님! 수정해주신 부분 모두 확인했어요 :)

첫 주부터 고생 많으셨습니다!
이만 머지할게요!
